distributed/protocol/core.py

```diff
@@ -0,0 +1,195 @@
+from __future__ import print_function, division, absolute_import
+
+from copy import deepcopy
+from functools import partial
+import logging
+
+try:
+    import pandas.msgpack as msgpack
+except ImportError:
+    import msgpack
+
+from toolz import identity, get_in, valmap
+
+from .compression import compressions, maybe_compress
+from .serialize import (serialize, deserialize, Serialize, Serialized,
+        to_serialize)
+from .utils import frame_split_size
+
+from ..utils import ignoring
+
+_deserialize = deserialize
+
+
+logger = logging.getLogger(__file__)
+
+
+def dumps(msg):
+    """ Transform Python message to bytestream suitable for communication """
+    try:
+        data = {}
+        # Only lists and dicts can contain serialized values
+        if isinstance(msg, (list, dict)):
+            msg, data = extract_serialize(msg)
+        small_header, small_payload = dumps_msgpack(msg)
+
+        if not data:  # fast path without serialized data
+            return small_header, small_payload
+
+        pre = {key: (value.header, value.frames)
+               for key, value in data.items()
+               if type(value) is Serialized}
+
+        data = {key: serialize(value.data)
+                     for key, value in data.items()
+                     if type(value) is Serialize}
+
+        header = {'headers': {},
+                  'keys': []}
+        out_frames = []
+
+        for key, (head, frames) in data.items():
+            if 'compression' not in head:
+                frames = frame_split_size(frames)
+                compression, frames = zip(*map(maybe_compress, frames))
+                head['compression'] = compression
+            head['lengths'] = list(map(len, frames))
+            header['headers'][key] = head
+            header['keys'].append(key)
+            out_frames.extend(frames)
+
+        for key, (head, frames) in pre.items():
```

Comment:
This is now tested in https://github.com/dask/distributed/pull/606/commits/9d688472c09cf84a83af0f0ba0f68454dcbf2ece